components/auth/hooks.ts

```diff
@@ -0,0 +1,100 @@
+import { FirebaseError } from "firebase/app"
+import {
+  AuthProvider,
+  createUserWithEmailAndPassword,
+  sendPasswordResetEmail,
+  signInWithEmailAndPassword,
+  signInWithPopup
+} from "firebase/auth"
+import { useAsyncCallback } from "react-async-hook"
+import { setProfile } from "../db"
+import { auth } from "../firebase"
+
+const errorMessages: Record<string, string | undefined> = {
+  // TODO: do we want to tell users this? this opens us up to user enumeration attacks
+  "auth/email-already-exists": "You already have an account.",
+  // same here
+  "auth/wrong-password": "Your password is wrong.",
+  "auth/invalid-email": "The email you provided is not a valid email.",
+  "auth/user-not-found": "You don't have an account."
+}
+
+const getErrorMessage = (errorCode?: string) => {
+  const niceErrorMessage = errorCode ? errorMessages[errorCode] : undefined
+  return niceErrorMessage || "Something went wrong!"
+}
+
+function useFirebaseFunction<Params, Result>(
+  fn: (params: Params) => Promise<Result>
+) {
+  return useAsyncCallback(async (params: Params) => {
+    try {
+      // necessary to await here so we trap any errors thrown by the function
+      const result = await fn(params)
+      return result
+    } catch (err) {
+      console.log(err)
+
+      const message = getErrorMessage(
+        err instanceof FirebaseError ? err.code : undefined
+      )
+      throw new Error(message)
+    }
+  })
+}
+
+export type CreateUserWithEmailAndPasswordData = {
+  email: string
+  fullName: string
+  nickname: string
+  password: string
+  confirmedPassword: string
+}
+
+export function useCreateUserWithEmailAndPassword() {
+  return useFirebaseFunction(
+    async ({
+      email,
+      fullName,
+      nickname,
+      password
+    }: CreateUserWithEmailAndPasswordData) => {
+      const credentials = await createUserWithEmailAndPassword(
+        auth,
+        email,
+        password
+      )
+
+      // TODO: handle errors here - potentially try to delete the created user to give a feel of "transactionality"?
+      await setProfile(credentials.user.uid, {
+        displayName: nickname,
+        fullName
+      })
```

Comment:
I tried using `useProfile` here, but the `uid` wasn't updating with the logged in user before reaching this call, so I added this function to set multiple properties on the profile in bulk.

Regardless, there's a possibility that a user is created, but this call to `setProfile` fails. Should we try to delete the new user and have the user restart the sign up process? Or?

Reply:
I think this works well! If we create the user but fail to update the profile, they'll figure out to log in, then they can set this in their profile.

If we want more control we could create a signup cloud function that would set these and return an auth token. We will likely create a fn for editing the profile, to ensure we update the display name everywhere.